File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_len = 0, i = 0;
        map<int,int> m;
        for(int j=0; j<s.size(); j++) {
            if(m.find(s[j]) != m.end() && m[s[j]] >= i) {
                i = m[s[j]] + 1;
            }
            m[s[j]] = j;
            max_len = max(max_len,j-i+1);
        }
        return max_len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> map = new HashMap<>();
        int n = s.length();
        int i = 0, j = 0;
        int max_len = 0;
        while(j < n) {
            if(map.containsKey(s.charAt(j))) {
                i = Math.max(i, map.get(s.charAt(j)) + 1);
            }
            map.put(s.charAt(j),j);
            max_len = Math.max(j-i+1,max_len);
            j++;
        }
        return max_len;
    }
}